### PR TITLE
[Transaction] Fix endless writing txn marker requests when the topic is deleted

### LINKED ISSUE #1386: [BUG] Endless WRITE_TXN_MARKERS requests even after producer is closed
**To Reproduce**

Run following application against a KoP standalone:

```java
        final var bootstrapServers = "localhost:9092";
        final var topic = "my-topic";

        @Cleanup final var admin = AdminClient.create(Collections.singletonMap(
                AdminClientConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapServers));

        admin.createTopics(Collections.singleton(new NewTopic(topic, 1, (short) 1))).all().get();

        final var props = new Properties();
        props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapServers);
        props.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
        props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
        props.put(ProducerConfig.TRANSACTIONAL_ID_CONFIG, "X0");
        final var producer = new KafkaProducer<String, String>(props);
        producer.initTransactions();
        producer.beginTransaction();
        producer.send(new ProducerRecord<>(topic, "hello"));
        producer.commitTransaction();
        producer.close();

        admin.deleteTopics(Collections.singleton(topic)).all().get();
```

- Kafka Client: 3.1.0
- Pulsar: 2.10.0
- KoP: 2.10.1.1

KoP configs:

```properties
messagingProtocols=kafka
protocolHandlerDirectory=./protocols
allowAutoTopicCreationType=partitioned
kafkaListeners=PLAINTEXT://127.0.0.1:9092
brokerEntryMetadataInterceptors=org.apache.pulsar.common.intercept.AppendIndexMetadataInterceptor
brokerDeleteInactiveTopicsEnabled=false
brokerDeleteInactivePartitionedTopicMetadataEnabled=true
brokerDeduplicationEnabled=true
kafkaTransactionCoordinatorEnabled=true
brokerDeleteInactiveTopicsMode=delete_when_subscriptions_caught_up
```

**Expected behavior**

**Additional context**

[completed-logs.log](https://github.com/streamnative/kop/files/9052910/pulsar-standalone-xuyunzedeMacBook-Pro.local.log)

### PR DESCRIPTION
**Pull request by Demogorgon314**
Fixes #1386 

### Motivation

Since the transaction operation will write into the transaction log first, 
the marker will write after the response to the client, 
however, when we commit the transaction and delete the topic Immediately, 
the marker write operation will always fail, 
so the write operation will always retry.

We need to check the topic exists when we try to write the marker.

### Modifications

* Add some debug log for writing marker log.
* Reusing publish message method.
* Add unit test for `TransactionMarkerChannelManager `.

### Documentation

Check the box below.

Need to update docs? 

- [ ] `doc-required` 
  
  (If you need help on updating docs, create a doc issue)
  
- [x] `no-need-doc` 
  
  (Please explain why)
  
- [ ] `doc` 
  
  (If this PR contains doc changes)

### CODE REVIEW

**Comment by BewareMyPower** [comment]
Could you update the PR description?

**Comment by Demogorgon314** [comment]
@BewareMyPower I have updated the description, PTAL.

File: kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/transaction/TransactionMarkerChannelManager.java [hunk 1]
```diff
@@ -472,7 +530,7 @@ private void drainQueuedTransactionMarkers() {
         }
     }
 
-    private synchronized void ensureDrainQueuedTransactionMarkersActivity() {
+    protected synchronized void ensureDrainQueuedTransactionMarkersActivity() {
```
Comment (BewareMyPower) [inline kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/transaction/TransactionMarkerChannelManager.java:533]:
```suggestion
    private synchronized void ensureDrainQueuedTransactionMarkersActivity() {
```

Keep it private.

**Comment by BewareMyPower** [comment]
The review is not done, I'll continue reviewing soon.

File: kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/storage/ReplicaManager.java [hunk 2]
```diff
@@ -111,6 +111,7 @@ public CompletableFuture<Map<TopicPartition, ProduceResponse.PartitionResponse>>
                         .thenAccept(offset -> addPartitionResponse.accept(topicPartition,
                                 new ProduceResponse.PartitionResponse(Errors.NONE, offset, -1L, -1L)))
                         .exceptionally(ex -> {
+                            log.error("Unexpected error when append records.", ex);
```
Comment (BewareMyPower) [inline kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/storage/ReplicaManager.java:114]: It's not an unexpected error unless `Errors.forException(ex.getCause())` returns `UNKNOWN_SERVER_ERROR`. This error is usually `NOT_LEADER_FOR_PARTITION` that trigger the retry of clients, in this case, we should not print the full stack.

File: tests/src/test/java/io/streamnative/pulsar/handlers/kop/coordinator/transaction/TransactionMarkerChannelManagerTest.java [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    private TransactionStateManager txnStateManager;
+
+    private OrderedScheduler scheduler;
+    private final MockTime time = new MockTime();
+    private final Set<TopicPartition> partitions = Sets.newHashSet(new TopicPartition("topic1", 0));
+    private final String transactionalId = "known";
+    private final long producerId = 10L;
+    private final short producerEpoch = 1;
+    private final int txnTimeoutMs = 1;
+    private final int coordinatorEpoch = 0;
```
Comment (BewareMyPower) [inline tests/src/test/java/io/streamnative/pulsar/handlers/kop/coordinator/transaction/TransactionMarkerChannelManagerTest.java:56]: Change these fields to local variables.

File: tests/src/test/java/io/streamnative/pulsar/handlers/kop/coordinator/transaction/TransactionMarkerChannelManagerTest.java [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+                TransactionResult.COMMIT,
+                txnMetadata,
+                transition,
+                "public/default"
+                );
+
+        assertEquals(transactionMarkerChannelManager.getTransactionsWithPendingMarkers().isEmpty(), !isTopicExists);
+    }
+
+}
```
Comment (BewareMyPower) [inline tests/src/test/java/io/streamnative/pulsar/handlers/kop/coordinator/transaction/TransactionMarkerChannelManagerTest.java:133]:
```suggestion
}

```

**Review by BewareMyPower — APPROVED** [review]

**Comment by BewareMyPower** [comment]
Could you migrate this PR to branch-2.8.3? There are some conflicts when I tried to cherry-pick it to branch-2.8.3.

**Comment by BewareMyPower** [comment]
BTW, https://github.com/streamnative/kop/pull/1399 also relies on this PR.

File: kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/transaction/TransactionMarkerChannelManager.java [hunk 5]
```diff
@@ -268,32 +274,80 @@ public void addTxnMarkersToBrokerQueue(String transactionalId,
...
-                addressAndPartitionMap.compute(address.get(), (__, set) -> {
-                    if (set == null) {
-                        set = new ArrayList<>();
-                    }
-                    set.add(topicPartition);
-                    return set;
-                });
-                addFuture.complete(null);
-            });
+            kopBrokerLookupManager.isTopicExists(pulsarTopic)
```
Comment (eolivelli) [inline kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/transaction/TransactionMarkerChannelManager.java:279]: isn't this check too heavy ?
we are on the write path here

Comment (Demogorgon314) [reply kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/transaction/TransactionMarkerChannelManager.java:279]: Yes, I also find out. Maybe we can get check if the topic exists for the cache?
```java
org.apache.pulsar.broker.service.BrokerService#fetchPartitionedTopicMetadataAsync
```

Comment (eolivelli) [reply kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/transaction/TransactionMarkerChannelManager.java:279]: @Demogorgon314 
Makes sense to me

**Pull request merged**